default/AI/ResourcesAI.py

```diff
@@ -29,98 +38,261 @@
-        if topPriority < 0:
-            topPriority = evaluationPair[0]
-        print "    ResourcePriority |Score: %s | %s " % (evaluationPair[0], evaluationPair[1])
-
-    # what is the focus of available resource centers?
-    print
-    warnings = {}
-    print "Planet Resources Foci:"
-    for planetID in empirePlanetIDs:
-        planet = universe.getPlanet(planetID)
-        planetPopulation = planet.currentMeterValue(fo.meterType.population)
-        maxPop = planet.currentMeterValue(fo.meterType.targetPopulation)
-        if maxPop < 1 and planetPopulation > 0:
-            warnings[planet.name] = (planetPopulation, maxPop)
-        statusStr = "  ID: " + str(planetID) + " Name: % 18s -- % 6s % 8s " % (str(planet.name), str(planet.size), str(planet.type))
-        statusStr += " Focus: % 8s" % ("_".join(str(planet.focus).split("_")[1:])[:8]) + " Species: " + str(planet.speciesName) + " Pop: %2d/%2d" % (planetPopulation, maxPop)
-        print statusStr
-    print "\n\nEmpire Totals:\nPopulation: %5d \nProduction: %5d\nResearch: %5d\n" % (empire.population(), empire.productionPoints, empire.resourceProduction(fo.resourceType.research))
-    if warnings != {}:
-        for pname in warnings:
-            mp, cp = warnings[pname]
-            print "Population Warning! -- %s has unsustainable current pop %d -- target %d" % (pname, cp, mp)
+        self.possible_output[self.current_focus] = (itarget, rtarget)
+        self.future_focus = self.current_focus
+
+
+class PlanetFocusManager(object):
+    """PlanetFocusManager tracks all of the empire's planets, what their current and future focus will be."""
+
+    def __init__(self):
+        universe = fo.getUniverse()
+
+        resource_timer.start("getPlanets")
+        planet_ids = list(PlanetUtilsAI.get_owned_planets_by_empire(universe.planetIDs))
+
+        # resource_timer.start("Shuffle")
+        # shuffle(generalPlanetIDs)
+
+        resource_timer.start("Targets")
+        planet_infos = [PlanetFocusInfo(universe.getPlanet(pid)) for pid in  planet_ids]
+        self.all_planet_info = {}
+        self.all_planet_info.update(zip(planet_ids, planet_infos))
+
+        self.raw_planet_info = dict(self.all_planet_info)
+        self.baked_planet_info = dict()
+
+        for pid, pinfo in self.raw_planet_info.items():
+            focusable = pinfo.planet.currentMeterValue(fo.meterType.targetPopulation) > 0
+            if not focusable:
+                self.baked_planet_info[pid] = self.raw_planet_info.pop(pid)
+
+
+    def bake_future_focus(self, pid, focus, update=True):
+        """Set the focus and moves it from the raw list to the baked list of planets
+        Return success or failure"""
+        pinfo = self.raw_planet_info.get(pid)
+        success = (pinfo != None and
+                   (pinfo.current_focus == focus
+                    or (focus in pinfo.planet.availableFoci
+                        and fo.issueChangeFocusOrder(pid, focus) == 1)))
```

Comment:
`fo.issueChangeFocusOrder(pid, focus)` return int. But in python you can treat it as bool. 

`and fo.issueChangeFocusOrder(pid, focus) == 1` => `and fo.issueChangeFocusOrder(pid, focus)`